Expression.parse bug fixes and support for underscores in numbers

Tomas-Domas opened the pull request:
- Fix #88121, Fix #81895 
  - The parser allowed nonsense expressions without throwing an error, such as `a a` or `5 a`
  - Modify the parsing process to disallow consecutive identifiers or constants.
- Resolve #73816, Resolve #97713, Resolve #84131
  - Ternary operations (i.e. if else) are not supported by the Expression class. However, trying to use them anyway is undefined behavior and only "works" because of the previously mentioned parser bug.
  - Modify the tokenization process such that "if" and "else" identifiers cause an error, which is valuable feedback to users.
- Resolve #84298
  - Underscores in numbers would get tokenized incorrectly, causing undefined behavior. For example, the number "1_000" would get tokenized as [constant of value `1`,   identifier of name `_000`], which is a nonsensical expression (think `5 a`).
  - Modify the tokenization process such that underscores are ignored in a number. (i.e. `1000` == `1_0_00`, `0xAB` == `0xA_B`).

*Bugsquad Edit:*
- *Fixes https://github.com/godotengine/godot/issues/97324*

HolonProduction commented:
Your changes are split across 6 commits, you will need to [squash](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) them into 1.

I also noticed that this PR does not contain [test cases](https://github.com/godotengine/godot/blob/master/tests/core/math/test_expression.h) for the linked issues, it would be great if you could add some based on the other tests from the file (there already seems to be a test case for underscore number literals, but it only checks the returned error and not the parse result so that would need to be extended).

Tomas-Domas commented:
> Your changes are split across 6 commits, you will need to [squash](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) them into 1.
> 
> I also noticed that this PR does not contain [test cases](https://github.com/godotengine/godot/blob/master/tests/core/math/test_expression.h) for the linked issues, it would be great if you could add some based on the other tests from the file (there already seems to be a test case for underscore number literals, but it only checks the returned error and not the parse result so that would need to be extended).

Ok thank you for the feedback! I will get to work on that next week.

Tomas-Domas committed:
Fix Expression bugs, add some features

Fix several bugs related to returning OK on parsing invalid expressions.
(ternary operators, missing operators, etc.)

Add support for underscores in numeric literals.

Tomas-Domas commented:
Thank you so much for all of your patience! I am still learning how to use git and how to manage the PR policies and whatnot, but I think I figured it out now. Working on the unit tests now!

HolonProduction commented:
Your PR seems to fail some of the existing test cases as well. Some stuff in there needs to be fixed, e.g. scientific number notation, some other failing cases seem somewhat questionable to me (e.g. `123 456` is tested to result in 123 but I'm inclined to say an error would be more intuitive) but since the expression behaviour isn't really well documented some core dev should take a look and make a decision on that.

Tomas-Domas commented:
> Your PR seems to fail some of the existing test cases as well. Some stuff in there needs to be fixed, e.g. scientific number notation, some other failing cases seem somewhat questionable to me (e.g. `123 456` is tested to result in 123 but I'm inclined to say an error would be more intuitive) but since the expression behaviour isn't really well documented some core dev should take a look and make a decision on that.

I just went through and added/modified several unit tests to have more consistent behavior. That one you mention resulted in `123` before too. I just modified it so it expects an error, which is what my code produces. 

It should be passing all tests now. Let me know if there's any other changes I should make!

gifford15 commented on core/math/expression.cpp:
```diff
@@ -1035,6 +1068,17 @@ Expression::ENode *Expression::_parse_expression() {
 				break;
+			case TK_IDENTIFIER:
+			case TK_CONSTANT:
+				// There's 2 consecutive constants/identifiers
```
I think adding error cases to the operator matching will be too narrow a fix. For instance, if an expression were `abc{foo}`, we should want that to fail to parse, but instead it will evaluate to the identifier `abc` because the `TK_CURLY_BRACKET_OPEN` token didn't trigger this safety. And the same for an Input token, and anything else that happens to not be matched as an operator or subsumed into the first matching expression. 

The root of the issue is that the code is not parsing greedily, but only as far as it requires to get an evaluable expression, and discarding the rest. So if it finds no operator to bring in an additional expression, it stops looking at any further input.
Rather than try to anticipate what may erroneously follow the expression that was parsed, a more robust and general solution would continue reading until the input was exhausted (`TK_EOF`) or any token was found out of place. Alternatively, where the code now assumes it is done once it completes an expression, it should check whether it consumed the entire input in order to produce it.
